
# Quality control scripts

**Table of contents:**

```table-of-contents
title: 
style: nestedList # TOC style (nestedList|nestedOrderedList|inlineFirstLevel)
minLevel: 2 # Include headings from the specified level
maxLevel: 6 # Include headings up to the specified level
includeLinks: true # Make headings clickable
hideWhenEmpty: false # Hide TOC if no headings are found
debugInConsole: false # Print debug info in Obsidian console
```

## Check strandedness from de novo assembly

- **Script**:  [`check_strand_incl_samtools.py`](../scripts/quality_control/check_strand_incl_samtools.py)
- **Description**: Check correct strandedness from RNA-seq data 
- **Dependencies**: subprocess, samtools
- **Tags**: #RNA-seq, #Alignment, #Strandedness, #BAM_processing, #Quality_control
- **Usage**: `python check_strand_incl_samtools.py -b File.bam -s R1_strand_info.txt -o strand_table.txt`
- **Input**: Bam file
- **Output**: 
	- Flag for each mapped R1 read (R1_strand_info.txt)
	- Reads on plus versus minus strand (and ratio) for all R1 reads (strand_table.txt)


## Extract sequence length and GC

- **Script**:  [`fasta_record_stats.py`](../scripts/quality_control/fasta_record_stats.py)
- **Description**: Calculate the length, GC content and nr of ambiguous bases for each record of a fasta file
- **Dependencies**: biopython
- **Tags**: #Quality_control, #FASTA
- **Usage**: `python fasta_record_stats.py -i data/genome.fna -o results/genome_stats.csv`
- **Input**: Nucleotide fasta file
- **Output**: Table with the record id, gc content, sequence length, and nr of ambiguous bp


## Get summary statistics for a numerical data column

- **Script**:  [`summarize_num_column.py`](../scripts/quality_control/summarize_num_column.py)
- **Description**: Script takes as input a tab-delimited file and the number of a numerical column to summarize. For the column of interest, the script outputs the mean, standard deviation as well as the 25th, 50th and 75th quantile range.
- **Dependencies**: pandas
- **Tags**: #summarize_data, #tabular_data
- **Usage**: `python summarize_num_column.py file.tab 3`
- **Input**: Tab-delimited file
- **Output**: Median, Mean, sd, 25,50,75th quantile 


## Parse Ribodetector Logfiles

- **Script**:  [`parse_ribodector.py`](../scripts/quality_control/parse_ribodector.py)
- **Description**: Software to parse the log files generated by ribodetector and returning the sample_name, total_sequences, rRNA_sequences and rRNA_percentage
- **Dependencies**: Pandas
- **Tags**: #Quality_control , #RNA-seq , #Log-parsing
- **Usage**: `python parse_ribodetector.py -i 05_quality_filtering/ribodetector -o 05_quality_filtering/ribodetector/summary.csv`
- **Input**: Folder with the ribodetector log files
- **Output**: Table with summary statistics
- **Related Snippets**:


## Count annotated splice junctions

- **Script**:  [`count_unannotated_SJ_file.sh`](../scripts/quality_control/count_unannotated_SJ_file.sh)
- **Description**: Script takes a list of STAR SJ files and counts how many of the junctions are annotated and unannotated
- **Dependencies**: 
- **Tags**: #Quality_control, #alignment, #splicing
- **Usage**: `count_junctions.sh "06_mapping/star/mapping/*_SJ.out.tab"`
- **Input**: List of Star SJ files
- **Output**: Summary table with the file name, total junctions, unannotated as well as annotated junctions
- **Related Snippets**:


## Parse the log files of FastP

- **Script**:  [`parse_fastp.py`](../scripts/quality_control/parse_fastp.py)
- **Description**: Parse the slurm log files when running FastP via slurm and output relevant summary statistics. In this example FastP was run with the script `../scripts/quality_control/run_fastp.sh`
- **Dependencies**: 
- **Tags**: #Log-parsing, #Quality_control 
- **Usage**: `python3 01_workflows_and_../scripts/parse_fastp.py -i "logs/fastp_54885_*.err" -o "04_read_quality/02_filtered_data/fastp_summary_v3.csv"`
- **Input**: Slurm log files
- **Output**: Summary statistics
- **Related Snippets**:


## Parse the log files generated by STAR

- **Script**:  [`parse_star.py`](../scripts/quality_control/parse_star.py)
- **Description**: Parse the log files from a STAR file and generate a file with summary statistics for several outputs
- **Dependencies**: Pandas
- **Tags**: #alignment , #Quality_control , #STAR
- **Usage**: `python parse_star.py --input_folder 06_mapping/star/mapping/ -o mapping_results.csv`
- **Input**: Several star log files
- **Output**: Table with STAR summary statistics
- **Related Snippets**:


## Estimate intron lengths from a gtf file

- **Script**:  [`estimate_introns_from_gtf.py`](../scripts/quality_control/estimate_introns_from_gtf.py)
- **Description**: Take a genome gtf file and extract intron sizes for each gene. Also outputs summary statistics, such as mean, median, min, max and 90/99/99.9 th length percentiles
- **Dependencies**: Pandas
- **Tags**: #gtf, #summarize_data , #genome 
- **Usage**: `python estimate_introns_from_gtf.py -i 03_data/Ochro1393_1_4_GeneCatalog_20181204.gtf -o 03_data/Ochromonas_gtf_introns.csv`
- **Input**: GTF
- **Output**: Table with intron information
- **Related Snippets**:


## Find potential pseudo-genes using the prokka output

- **Script**:  [`prokka-suggest_pseudogenes.pl`](../scripts/quality_control/prokka-suggest_pseudogenes.pl)
- **Description**: Script to parse prokka output and number potential pseudogenes (originally from [https://raw.githubusercontent.com/tseemann/bioinfo-../scripts/master/bin/prokka-suggest_pseudogenes.pl](https://raw.githubusercontent.com/tseemann/bioinfo-../scripts/master/bin/prokka-suggest_pseudogenes.pl))
- **Dependencies**: 
- **Tags**: #Pseudogene, #Prokka
- **Usage**: `perl prokka-suggest_pseudogenes.pl prokka.faa`
- **Input**: Prokka faa file
- **Output**: Number of potential pseudogenes
- **Related Snippets**:


## Summarize summary statistics table from seqkit

- **Script**:  [`summarize_seqkit.py`](../scripts/quality_control/summarize_seqkit.py)
- **Description**: Get sum, mean, median, min, max for seqkit summary output
- **Dependencies**: pandas, tabulate
- **Tags**: #Quality_control
- **Usage**: `python scripts/summarize_seqkit.py -i results/seqkit/seqkit_stats.tsv -o results/seqkit/seqkit_stats_summary.tsv`
- **Input**: Txt file
- **Output**: 
	- File and md output for summary statistics